access/token.go

```diff
@@ -0,0 +1,299 @@
+	DefaultTokenTTL = time.Hour * 24 * 14
+)
+
+// Token abstracts a json web token
+type Token = jwt.Token
+
+// TokenClaims is a JWT Claims object
+type TokenClaims struct {
+	*jwt.StandardClaims
+	Username string `json:"username"`
+}
+
+// ParseToken will parse, validate and return a token
+func ParseToken(tokenString string, tokens TokenSource) (*Token, error) {
+	return jwt.Parse(tokenString, tokens.VerificationKey)
+}
+
+// TokenSource creates tokens, and provides a verification key for all tokens
+// it creates
+//
+// implementations of TokenSource must conform to the assertion test defined
+// in the spec subpackage
+type TokenSource interface {
+	CreateToken(pro *profile.Profile, ttl time.Duration) (string, error)
+	CreateTokenWithClaims(claims jwt.MapClaims, ttl time.Duration) (string, error)
+	// VerifyKey returns the verification key for a given token
+	VerificationKey(t *Token) (interface{}, error)
+}
+
+type pkTokenSource struct {
+	pk            crypto.PrivKey
+	signingMethod jwt.SigningMethod
+	verifyKey     *rsa.PublicKey
+	signKey       *rsa.PrivateKey
+}
+
+// assert pkTokenSource implements tokens at compile time
+var _ TokenSource = (*pkTokenSource)(nil)
+
+// NewPrivKeyTokenSource creates an authentication interface backed by a single
+// private key. Intended for a node running as remote, or providing a public API
+func NewPrivKeyTokenSource(privKey crypto.PrivKey) (TokenSource, error) {
+	methodStr := ""
+	keyType := privKey.Type().String()
+	switch keyType {
+	case "RSA":
+		methodStr = "RS256"
+	default:
+		return nil, fmt.Errorf("unsupported key type for token creation: %q", keyType)
+	}
+
+	signingMethod := jwt.GetSigningMethod(methodStr)
+
+	rawPrivBytes, err := privKey.Raw()
+	if err != nil {
+		return nil, err
+	}
+	signKey, err := x509.ParsePKCS1PrivateKey(rawPrivBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	rawPubBytes, err := privKey.GetPublic().Raw()
+	if err != nil {
+		return nil, err
+	}
+	verifyKeyiface, err := x509.ParsePKIXPublicKey(rawPubBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	verifyKey, ok := verifyKeyiface.(*rsa.PublicKey)
+	if !ok {
+		return nil, fmt.Errorf("public key is not an RSA key. got type: %T", verifyKeyiface)
+	}
+
+	return &pkTokenSource{
+		pk:            privKey,
+		signingMethod: signingMethod,
+		verifyKey:     verifyKey,
+		signKey:       signKey,
+	}, nil
+}
+
+// CreateToken returns a new JWT token
+func (a *pkTokenSource) CreateToken(pro *profile.Profile, ttl time.Duration) (string, error) {
+	// create a signer for rsa 256
+	t := jwt.New(a.signingMethod)
+
+	var exp int64
+	if ttl != time.Duration(0) {
+		exp = Timestamp().Add(ttl).In(time.UTC).Unix()
+	}
+
+	// set our claims
+	t.Claims = &TokenClaims{
+		StandardClaims: &jwt.StandardClaims{
+			Subject: pro.ID.String(),
+			// set the expire time
+			// see http://tools.ietf.org/html/draft-ietf-oauth-json-web-token-20#section-4.1.4
+			ExpiresAt: exp,
+		},
+		Username: pro.Peername,
```

Comment:
This specific PR is less concerned with the implementation specifics and more on raw mechanics. Also this will evolve a bit as we deepen the integration with token based out between systems. 

With that said, this particular spot does push the notion of a username, however all the integration work done so far only has it included for convenience reasons and is not used as a real reference to anything. A lot more detail on the more real use cases can be seen in [this integration](https://github.com/qri-io/cloud/pull/433). 

As such, I expect this to go away or be modified to suit a more general usecase.